File: clients/libjabi/peripherals/spi.cpp
#include <cstring>
#include <libjabi/byteorder.h>
#include <libjabi/interfaces/interface.h>

namespace jabi {

#include <jabi/peripherals.h>
#include <jabi/peripherals/spi.h>

void Device::spi_set_freq(int freq, int idx) {
    iface_dynamic_req_t req = {
        .msg = {
            .periph_id   = PERIPH_SPI_ID,
            .periph_idx  = static_cast<uint16_t>(idx),
            .periph_fn   = SPI_SET_FREQ_ID,
            .payload_len = sizeof(spi_set_freq_req_t),
            .payload     = {0},
        },
        .payload = std::vector<uint8_t>(sizeof(spi_set_freq_req_t), 0),
    };

    auto args = reinterpret_cast<spi_set_freq_req_t*>(req.payload.data());
    args->freq = htole<uint32_t>(freq);

    iface_dynamic_resp_t resp = interface->send_request(req);
    if (resp.payload.size() != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

void Device::spi_set_mode(int mode, int idx) {
    iface_dynamic_req_t req = {
        .msg = {
            .periph_id   = PERIPH_SPI_ID,
            .periph_idx  = static_cast<uint16_t>(idx),
            .periph_fn   = SPI_SET_MODE_ID,
            .payload_len = sizeof(spi_set_mode_req_t),
            .payload     = {0},
        },
        .payload = std::vector<uint8_t>(sizeof(spi_set_mode_req_t), 0),
    };

    auto args = reinterpret_cast<spi_set_mode_req_t*>(req.payload.data());
    args->mode = static_cast<uint8_t>(mode);

    iface_dynamic_resp_t resp = interface->send_request(req);
    if (resp.payload.size() != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

void Device::spi_set_bitorder(bool msb, int idx) {
    iface_dynamic_req_t req = {
        .msg = {
            .periph_id   = PERIPH_SPI_ID,
            .periph_idx  = static_cast<uint16_t>(idx),
            .periph_fn   = SPI_SET_BITORDER_ID,
            .payload_len = sizeof(spi_set_bitorder_req_t),
            .payload     = {0},
        },
        .payload = std::vector<uint8_t>(sizeof(spi_set_bitorder_req_t), 0),
    };

    auto args = reinterpret_cast<spi_set_bitorder_req_t*>(req.payload.data());
    args->order = msb;

    iface_dynamic_resp_t resp = interface->send_request(req);
    if (resp.payload.size() != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

void Device::spi_write(std::vector<uint8_t> data, int idx) {
    iface_dynamic_req_t req = {
        .msg = {
            .periph_id   = PERIPH_SPI_ID,
            .periph_idx  = static_cast<uint16_t>(idx),
            .periph_fn   = SPI_WRITE_ID,
            .payload_len = static_cast<uint16_t>(data.size()),
            .payload     = {0},
        },
        .payload = data,
    };

    iface_dynamic_resp_t resp = interface->send_request(req);
    if (resp.payload.size() != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

std::vector<uint8_t> Device::spi_read(size_t len, int idx) {
    iface_dynamic_req_t req = {
        .msg = {
            .periph_id   = PERIPH_SPI_ID,
            .periph_idx  = static_cast<uint16_t>(idx),
            .periph_fn   = SPI_READ_ID,
            .payload_len = sizeof(spi_read_j_req_t),
            .payload     = {0},
        },
        .payload = std::vector<uint8_t>(sizeof(spi_read_j_req_t)),
    };

    auto args = reinterpret_cast<spi_read_j_req_t*>(req.payload.data());
    args->data_len = htole<uint16_t>(static_cast<uint16_t>(len));

    iface_dynamic_resp_t resp = interface->send_request(req);
    if (resp.payload.size() != len) {
        throw std::runtime_error("unexpected payload length");
    }
    return resp.payload;
}

std::vector<uint8_t> Device::spi_transceive(std::vector<uint8_t> data, int idx) {
    iface_dynamic_req_t req = {
        .msg = {
            .periph_id   = PERIPH_SPI_ID,
            .periph_idx  = static_cast<uint16_t>(idx),
            .periph_fn   = SPI_TRANSCEIVE_ID,
            .payload_len = static_cast<uint16_t>(data.size()),
            .payload     = {0},
        },
        .payload = data,
    };

    iface_dynamic_resp_t resp = interface->send_request(req);
    if (resp.payload.size() != data.size()) {
        throw std::runtime_error("unexpected payload length");
    }
    return resp.payload;
}

};

File: clients/jabi-rs/src/peripherals/spi.rs
use crate::interfaces::InterfaceRequest;
use crate::Error;
use deku::{DekuContainerWrite, DekuUpdate, DekuWrite};

enum Func {
    SetFreq = 0,
    SetMode = 1,
    SetBitorder = 2,
    Write = 3,
    Read = 4,
    Transceive = 5,
}

fn gen_req(f: Func, idx: usize, payload: Vec<u8>) -> Result<InterfaceRequest, Error> {
    let mut r = InterfaceRequest {
        periph_id: crate::InstID::SPI as u16,
        periph_idx: idx as u16,
        periph_fn: f as u16,
        payload_len: 0,
        payload: payload,
    };
    r.update().map_or(Err(Error::PacketFormat), |_| Ok(r))
}

impl crate::Device {
    pub fn spi_set_freq(&self, idx: usize, freq: u32) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct SetFreqRequest {
            freq: u32,
        }
        let req = SetFreqRequest { freq };
        let resp = self.send(&gen_req(Func::SetFreq, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn spi_set_mode(&self, idx: usize, mode: u8) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct SetModeRequest {
            mode: u8,
        }
        let req = SetModeRequest { mode };
        let resp = self.send(&gen_req(Func::SetMode, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn spi_set_bitorder(&self, idx: usize, msb: bool) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct SetBitorderRequest {
            order: u8,
        }
        let req = SetBitorderRequest { order: msb as u8 };
        let resp = self.send(&gen_req(Func::SetBitorder, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn spi_write(&self, idx: usize, data: &[u8]) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct WriteRequest<'a> {
            data: &'a [u8],
        }
        let req = WriteRequest { data };
        let resp = self.send(&gen_req(Func::Write, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn spi_read(&self, idx: usize, len: usize) -> Result<Vec<u8>, Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct ReadRequest {
            data_len: u16,
        }
        let req = ReadRequest {
            data_len: len as u16,
        };
        let resp = self.send(&gen_req(Func::Read, idx, req.to_bytes().unwrap())?)?;
        Ok(resp)
    }

    pub fn spi_transceive(&self, idx: usize, data: &[u8]) -> Result<Vec<u8>, Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct TransceiveRequest<'a> {
            data: &'a [u8],
        }
        let req = TransceiveRequest { data };
        let resp = self.send(&gen_req(Func::Transceive, idx, req.to_bytes().unwrap())?)?;
        Ok(resp)
    }
}
